### DISCUSSION TITLE
Code bias issue

### DISCUSSION THREAD

**Discussion by ParisaShafiei** [groops-devs/groops#190] [Q&A]
Hi @tmayerguerr, 

I have found something strange related to both parametrization and estimation of receiver code biases. I am not sure if this is an issue or not, but here it is:

For the kinematic orbit processing the CubeSats, I added these parametrziations as setup: 
ionosphereSTEC
ionosphereVTEC 
clocks
kinematicPositions
signalBiases
codeBiases
tecBiases
ambiguities
receiverAntenna 
trasnmitterAntenna

As you see, I have defined the codeBiases in my parametrization, but in the first part of the log file in the terminal (please see below), no parameters are set up for the code biases and to my surprise, at the end of the processing, the code biases are estimated in the "signalBias.text" file. 

2023-05-05 14:13:34 Status  (process  8) setup parameters
2023-05-05 14:13:34 Info    (process  8)     53769 constrained STEC parameters (preeliminated)
2023-05-05 14:13:34 Info    (process  8)      7138 VTEC epoch parameters
2023-05-05 14:13:34 Info    (process  8)      7138 receiver clock error parameters
2023-05-05 14:13:34 Info    (process  8)     21414 kinematic position parameters
2023-05-05 14:13:34 Info    (process  8)         1 receiver TEC bias parameters
2023-05-05 14:13:34 Info    (process  8)         2 receiver phase bias parameters
2023-05-05 14:13:34 Info    (process  8)      2650 ambiguity parameters
2023-05-05 14:13:34 Info    (process  8) + =======
2023-05-05 14:13:34 Info    (process  8)     38343 parameters in 7181 normal equation matrix blocks

For processing other LEO satellites, I have not had this problem. For example for GRACE-FO, the code biases are in the parametrization setup, appeared in the summary of processing and estimated at the end, but in my case, it is completely different. 

I am wondering why although they are not shown in the setup of parametrization in the terminal, they are still estimated at the end of processing! 
Can it be because of the high values of the estimated code biases? (please see these values below)

type   bias [m]
===============================
 C1CG** -3.523668285826964564e+01
 C2LG** -5.803285907407789779e+01
 L1CG** -7.384220400430327458e-02
 L2LG** -6.830510536681960732e-03

Thank you in advance for your time and support. It is really appreciated. 

I am looking forward to hearing from you.

With kind regards, 
Parisa

**Comment by zhedumi** [comment]
Hi Parisa,

that seems somewhat weird that it does not show it. Could it be that you might read some biases in and forgot to add it int he parametrization to enable the code biases specifically? I currently cannot think of a reason why it would not show the code biases otherwise. 

best regards

Patrick

**Reply by ParisaShafiei** [reply]
Hi @zhedumi, 

Good morning :)

Thanks for your reply. Yes for me it is also strange why it does not show the code bias parameters although I have them in my parametrizations. 

And this is only for these satellites, because when I used the example **03groopsGnssProcessing.xml** of GRACE FO for processing  GRACE data, I did not have this problem, but when I adopted it  for processing these CubeSats,  I faced this issue. 

Here there is a screenshot showing the parametrizations in my config file.

![image](https://github.com/groops-devs/groops/assets/99748448/3e1d8d0e-f115-457e-a1b6-f210a1a5c1f2)



If you would like to check the config file, please let me know. I can share it with you. 

Looking forward to hearing from you...

With kind regards, 
Parisa

**Comment by zhedumi** [comment]
Hi,

I think I know the issue. The issue is that you do not have enough boservations to actually estimate the biases. The outputlog does not tell you but incase you lack the number of observations to estimate the bias correctly it iwll just not do it. And that is the case since you only have 2 code observation types but estimate tec bias, code bias etc. So it will just ignore the set parameter. 

Best regards,
Patrick

**Reply by ParisaShafiei** [reply]
Hi Patrick, 

Thanks for your reply. 

I agree with you... I found the same issue when I processed an IGS station, i.e. ZIMM in PPP.  This what I did:

First, I used the RINEX 2 version and launch the processing. 
Types of observations: C1 L1 S1 C2 L2 S2 P2 C5 L5 S5
In this case, no code bias parameters were considered in parametrization, but I got C1CG** and C2WG** code biases in the signalBias output file.

Secondly, I used the RINEX3 version in which 2 more observation types were considered. 
Types of observations: C1C L1C S1C C2W L2W S2W C2X L2X S2X C5X L5X S5X
In this case, I got 2 receiver code bias parameters in parametrization and same as the first case, I got both C1CG** and C2WG** code biases  in the signalBias output file.

On the other hand for GRACE Follow On, we have these observation types: C1C, C1W, C2W (i.e. C/A, P1 and P2) and when I asked to ignore the C1W or C1C, it did not consider the code biases in parametrization like what happened for these CubeSats. 

So, based on what you said and what I got by testing, it means that for these CubeSats, since I only have C1C and  C2L code observation, it cannot compute the code biases properly which means two code observations are not enough...

Now, my question is can we trust the estimated code biases in this case? 

Sorry for the long explanation and thank you so much for your time and help :)

With kind regards, 
Parisa

**Reply by zhedumi** [reply]
Sorry for the late reply. 
Just need some clarification since I am a bit out of the loop. 
When we talk about the signalBiases we talk about the transmitter or the receiverSignalbiases? Cause the transmitterBiases will not be estimateable at all in PPP. So I presume the estimation for the transmitter is generally deactivated.
Also are both the transmitterSignalbiases and the receiverSignalbiases deactivated int he estimation? 

One thing you definitly should do is deactivate estimation of the transmitterbias and use maybe a provided file. Then there would be the issue of separating the TEC, Clock and receiverbiases which can only be done relatively to each other.  

What you mean exactly by trusting the code biases? They were not estimated in case of too few observations and therefore I would not trust them to be used in any further analysis since. But I will double check just in case.

Best regards,
Patrick

**Reply by ParisaShafiei** [reply]
Hi Patrick, 

Thanks for your reply.

Regarding your first question, it is about receiver signal biases that I try to estimate in my LEO POD processing. So, in my estimation, as you can see in the screenshot, I only asked to estimate the receiver signal biases. 

![image](https://github.com/groops-devs/groops/assets/99748448/657d7875-1888-4f5b-bc5c-5f2cc34dbb22)


So, as you remember I already processed the GNSS products in 2021 and I provided the estimated signal biases for transmitters in my processing. 

However, even though I only have two code observations, i.e. C1C and C2L, and as you said they are not enough for estimating the code biases, at the end of processing, I got the signal biases for the LEO CubeSats. It sounds strange to me and because of that I asked you should I trust these estimated biases or not....

I have attached one of the output signals biases so you can take a look at it. 

[signalBias_2021-08-01_LEO_receiver.txt](https://github.com/groops-devs/groops/files/12421437/signalBias_2021-08-01_LEO_receiver.txt)

With kind regards, 
Parisa

**Comment by FFFFFG-IGG** [comment]
> Hi @tmayerguerr,
> 
> I have found something strange related to both parametrization and estimation of receiver code biases. I am not sure if this is an issue or not, but here it is:
> 
> For the kinematic orbit processing the CubeSats, I added these parametrziations as setup: ionosphereSTEC ionosphereVTEC clocks kinematicPositions signalBiases codeBiases tecBiases ambiguities receiverAntenna trasnmitterAntenna
> 
> As you see, I have defined the codeBiases in my parametrization, but in the first part of the log file in the terminal (please see below), no parameters are set up for the code biases and to my surprise, at the end of the processing, the code biases are estimated in the "signalBias.text" file.
> 
> 2023-05-05 14:13:34 Status (process 8) setup parameters 2023-05-05 14:13:34 Info (process 8) 53769 constrained STEC parameters (preeliminated) 2023-05-05 14:13:34 Info (process 8) 7138 VTEC epoch parameters 2023-05-05 14:13:34 Info (process 8) 7138 receiver clock error parameters 2023-05-05 14:13:34 Info (process 8) 21414 kinematic position parameters 2023-05-05 14:13:34 Info (process 8) 1 receiver TEC bias parameters 2023-05-05 14:13:34 Info (process 8) 2 receiver phase bias parameters 2023-05-05 14:13:34 Info (process 8) 2650 ambiguity parameters 2023-05-05 14:13:34 Info (process 8) + ======= 2023-05-05 14:13:34 Info (process 8) 38343 parameters in 7181 normal equation matrix blocks
> 
> For processing other LEO satellites, I have not had this problem. For example for GRACE-FO, the code biases are in the parametrization setup, appeared in the summary of processing and estimated at the end, but in my case, it is completely different.
> 
> I am wondering why although they are not shown in the setup of parametrization in the terminal, they are still estimated at the end of processing! Can it be because of the high values of the estimated code biases? (please see these values below)
> 
> # type bias [m]
> C1CG** -3.523668285826964564e+01 C2LG** -5.803285907407789779e+01 L1CG** -7.384220400430327458e-02 L2LG** -6.830510536681960732e-03
> 
> Thank you in advance for your time and support. It is really appreciated.
> 
> I am looking forward to hearing from you.
> 
> With kind regards, Parisa

Hi Parisa,

  From sebasbian's PhD thesis my best guess for your question is that number of observation type is not enough to estimate the total code bias. As can be seen in section 6.5 of ‘Reprocessing Multiple GNSS Constellations and a Global Station Network from 1994 to 2020 with the Raw Observation Approach (sebasbian 2022)’, one can learn that the parameters of code bias are firstly divided into estimable parts and parts in none space using eigenvalue decomposition, the estimable code bias are firstly set in parametrization and then combined the code bias from non space that will be re-constructed from a global scale, the combination of estimable and none space parts are the final code bias as shown in your results. This means in short is that code bias of C1/C2 may in the none space for not adequate bands are provided, they will be not set and estimated straightly in parametrization part but will be reconstruted when combining the code bias from none space.

 In my opnion, receiver code bias are not a key influence on the final result when using PPP-AR products, as it can be different from various algorithms. If you want to know the reasionbility of the results, check the residuals of the parameters may be a good choice.

   That's all I've learned about this case so far, and you may get some professional instructions from the developers.

Regards.
Yongzhao Fan

**Reply by tmayerguerr** [reply]
Dear Yongzhao Fan,
Thanks for helping. You are completely right.
Torsten
